### PR TITLE
Caching on GHES

### PR DESCRIPTION
**Pull request by dmitry-shibanov**
**Description:**
In scope of this pull request we add support for caching on GHES.
**Related issue:**
https://github.com/actions/setup-node/issues/450

**Check list:**
- [ ] Mark if documentation changes are required.
- [ ] Mark if tests were added or updated to cover the changes.

### CODE REVIEW

**Comment by dmitry-shibanov** [comment]
Hello everyone. Could you please take a look at this pull request.
cc: @tiwarishub @brcrista

File: __tests__/cache-restore.test.ts [hunk 1]
```diff
@@ -102,6 +103,9 @@ describe('cache-restore', () => {
 
     // cache-utils
     getCommandOutputSpy = jest.spyOn(utils, 'getCommandOutput');
+
```
Comment (tiwarishub) [inline __tests__/cache-restore.test.ts:106]: Is this change and L53 change is required? Because I think in restore restoreCache function we don't check ``isCacheActionAvailable`` functions. This we are checking in main.ts

File: __tests__/cache-save.test.ts [hunk 2]
```diff
@@ -70,6 +71,9 @@ describe('run', () => {
 
     // utils
     getCommandOutputSpy = jest.spyOn(utils, 'getCommandOutput');
+
```
Comment (tiwarishub) [inline __tests__/cache-save.test.ts:74]: Same as above comment. Is this change required here

File: __tests__/cache-utils.test.ts [hunk 3]
```diff
@@ -51,7 +40,35 @@ describe('cache-utils', () => {
     });
   });
 
+  it('isCacheFeatureAvailable is false', () => {
```
Comment (tiwarishub) [inline __tests__/cache-utils.test.ts:43]: in the description, Please add whether the scenario is of GHES or not.

File: __tests__/cache-utils.test.ts [hunk 4]
```diff
@@ -51,7 +40,35 @@ describe('cache-utils', () => {
...
+    isFeatureAvailable.mockImplementation(() => false);
+    process.env['GITHUB_SERVER_URL'] = 'https://www.test.com';
+
+    expect(isCacheFeatureAvailable()).toBe(false);
+    expect(info).toHaveBeenCalledWith(
+      '[warning]Cache action is only supported on GHES version >= 3.5. If you are on version >=3.5 Please check with GHES admin if Actions cache service is enabled or not.'
+    );
+  });
+
+  it('isCacheFeatureAvailable is false', () => {
```
Comment (tiwarishub) [inline __tests__/cache-utils.test.ts:53]: Same here

File: src/cache-utils.ts [hunk 5]
```diff
@@ -95,3 +96,32 @@ export const getCacheDirectoryPath = async (
...
+  const ghUrl = new URL(
+    process.env['GITHUB_SERVER_URL'] || 'https://github.com'
+  );
+  return ghUrl.hostname.toUpperCase() !== 'GITHUB.COM';
+}
+
+export function isCacheFeatureAvailable(): boolean {
+  if (!cache.isFeatureAvailable()) {
+    if (isGhes()) {
+      logWarning(
```
Comment (tiwarishub) [inline src/cache-utils.ts:115]: I would prefer here to throw an error instead of a warning. Because this keeps it consistent with other setup-* actions https://github.com/actions/setup-java/pull/308/files#diff-3294a832ea2276e554177e0b3007cc2d401c082912c7fbde49fa09141bf1aed1R87-R103

File: src/main.ts [hunk 6]
```diff
@@ -46,7 +46,7 @@ export async function run() {
     }
 
     if (cache) {
-      if (isGhes()) {
+      if (!isCacheFeatureAvailable()) {
         throw new Error('Caching is not supported on GHES');
```
Comment (tiwarishub) [inline src/main.ts:50]: Let's not throw the error here (if we can throw it from isCacheFeatureAvailable function ). We can return if the Cache feature is not available

**Review by tiwarishub — APPROVED** [review]
Thank you for taking this PR 🙇 . I have left few comments

File: src/cache-utils.ts [hunk 7]
```diff
@@ -95,3 +96,28 @@ export const getCacheDirectoryPath = async (
...
+
+export function isCacheFeatureAvailable(): boolean {
+  if (!cache.isFeatureAvailable()) {
+    if (isGhes()) {
+      throw new Error(
+        'Cache action is only supported on GHES version >= 3.5. If you are on version >=3.5 Please check with GHES admin if Actions cache service is enabled or not.'
+      );
+    } else {
+      core.warning(
+        'An internal error has occurred in cache backend. Please check https://www.githubstatus.com/ for any ongoing issue in actions.'
```
Comment (brcrista) [inline src/cache-utils.ts:115]: @dmitry-shibanov in https://github.com/actions/setup-python/pull/363 we ended up changing this warning:

```suggestion
        'The runner was not able to contact the cache service. Caching will be skipped'
```

Comment (dmitry-shibanov) [reply src/cache-utils.ts]: Done. Thank you.

File: src/main.ts (hunk 6 above)
Comment (dmitry-shibanov) [reply src/main.ts]: Done. Thank you.

File: src/cache-utils.ts (hunk 5 above)
Comment (dmitry-shibanov) [reply src/cache-utils.ts]: Done. Thank you.

File: __tests__/cache-utils.test.ts (hunk 4 above)
Comment (dmitry-shibanov) [reply __tests__/cache-utils.test.ts]: Done. Thank you.

File: __tests__/cache-utils.test.ts (hunk 3 above)
Comment (dmitry-shibanov) [reply __tests__/cache-utils.test.ts]: Done. Thank you.

File: __tests__/cache-save.test.ts (hunk 2 above)
Comment (dmitry-shibanov) [reply __tests__/cache-save.test.ts]: Done. Thank you.

File: __tests__/cache-restore.test.ts (hunk 1 above)
Comment (dmitry-shibanov) [reply __tests__/cache-restore.test.ts]: Done. Thank you.

**Review by tiwarishub — APPROVED** [review]

**Review by marko-zivic-93 — APPROVED** [review]

**Review by vsafonkin — APPROVED** [review]
